[비동기 댓글 기능] 베디 미션 제출합니다

dpudpu opened the pull request:
제이슨 안녕하세요! 잘 부탁드립니다.

dpudpu committed:
refactor: 패키지 구조 mvc로 변경

dpudpu committed:
refactor: controller 패키지 구조 변경

dpudpu committed:
feat: 작성글 갯수 확인 api 구현

dpudpu committed:
feat: 댓글 작성 api 구현

dpudpu committed:
refactor: service layer, 다른 service 의존하게 변경

dpudpu committed:
refactor: ArticleServiceTest Mockito 로 구현

dpudpu committed:
feat: RollingFileAppender 추가

dpudpu committed:
refactor: CommentApiTest ObjectMapper 적용

dpudpu committed:
test: CommentServiceTest slicesTest 로 구현

dpudpu committed:
feat: CommentApi CRUD 추가

dpudpu committed:
feat: comments.js crud ajax 구현

dpudpu committed:
refactor: error.html 에러처리 방식 수정

dpudpu committed:
feat: logback-spring RollingFileAppender 추가

dpudpu committed:
feat: deploy.sh 쉘스크립트 추가

dpudpu committed:
feat: Logging 컬러 추가

dpudpu committed:
style: 포맷팅

woowahan-pjs reviewed (changes requested):
전체적인 설계와 깔끔한 코드 구현이 좋았습니다. 👍 
`AuthInterceptor`, `GuestInterceptor`, `LoginInterceptor`의 역할에 대해 설명해 주실 수 있나요?

dpudpu committed:
rename: ArticleApi, CommentApi -> ArticleRestController, CommentRestController

dpudpu committed:
refactor: ArticleRestController.getCount() 응답 HttpStatus 변경

dpudpu committed:
refactor: interceptor 패스 추가

comment 리스트 보기 로그인 안 해도 볼 수 있게 추가
path.equals("/api/articles/**/comments") && method.equals("GET");

dpudpu committed:
test: Mockito 학습테스트 추가

dpudpu commented:
### LoginInterceptor, GuestInterceptor

인터셉터가 원래는 하나였는데, 중간에 몇가지 조건`(로그인 상태에서는 로그인, 회원가입 페이지에 접근이 불가능하다)`이 더 추가가 되니까 인터셉터 코드가 더러워지더라고요. 

ex)
```java
        boolean isLogin = session.getAttribute(USER_SESSION) != null;
        if (isLogin) {
            return isNotException(response, path, method);
        }

        if (!(isLoginForm(path) || isSignup(path, method) || isSignupForm(path))) {
            response.sendRedirect("/users/login");
            return false;
        }
```

그래서 조금이라도 코드를 깔끔하게 관리하기 위해서 로그인 했을 때와 로그인을 하지 않았을 때의 인터셉터를 나누어줬습니다.



### AuthInterceptor 

`AuthInterceptor` 다른 사람의  마이페이지, 정보 수정을 못하게 하려면 ` /user/{id}`의 {id}와 로그인한 유저의 id 값을 비교해주는 작업을 해야하는데 여러 곳에서 검사가 필요하다보니 중복이 생기더라고요.

```java
@GetMapping("/{id}")
public String show(@PathVariable Long id,
                   Model model,
                   UserSession userSession) {
    /// 중복 되는 코드 
    if (!id.equals(userSession.getId())) {
        return "redirect:/";
    }
    ////
    
    UserDto.Response userResponseDto = userService.getUserById(id);
    model.addAttribute("user", userResponseDto);
    return "mypage";
}
```

그래서 인터셉터로 관리해보자 했는데 좋은 방법을 못찾아서 일단 임시로 AuthInterceptor를 만들었습니다. (임시방편)

일단 AuthInterceptor를 지우고 `identify` 메소드를 호출해서 검사하고 예외처리 해주는 방식으로 변경했습니다.

```java
 @DeleteMapping("/{id}")
    public String delete(@PathVariable Long id, HttpSession session, UserSession userSession) {
        identify(userSession, id);

        userService.deleteById(id);
        session.removeAttribute(USER_SESSION);
        return "redirect:/";
    }

    private void identify(UserSession userSession, Long id) {
        if (!id.equals(userSession.getId())) {
            throw new AuthException("본인이 아닙니다.");
        }
    }
```

dpudpu committed:
refactor: UserController 본인 확인 AuthInterceptor 가 아닌 Controller에서 하게 변경

dpudpu committed:
style: 필요없는 주석 제거

dpudpu committed:
refactor: deploy.sh

dpudpu commented:
추가 질문이 있습니다. 

현재 Service가 다른 Service를 의존하고 있는데 여기서 고민 되는 부분이 있습니다. 
Controller에서 Service에서 User를 조회할 때는 dto로 리턴을, 다른 Service에서 조회할 때는 엔티티로 리턴 해줍니다. 
그러다보니 비슷한 역할을 하는 거 같아서 이름도 헷갈리고 중복느낌이 납니다.

ex)
```java
    public UserDto.Response getUserById(Long id) {
        final User user = findById(id);

        return UserDto.Response.createByUser(user);
    }

    public User findById(final Long id) {
        return userRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("존재하지 않는 유저입니다. : " + id));
    }
```


그래서 구조를 
Controller -> Service -> InnerService -> Repository 이렇게 해볼까 하는데 어떻게 생각하시나요?
InnerService 역할은 Repository와 관련된 작업만 해주고 
Service 역할은 InnerServce에서 받은 엔티티로 도메인간의 연결, 매핑과 관련된 일들을 해줍니다.

ex)
```java
@Service
public class UserInnerService {
    private final UserRepository userRepository;

    public UserInnerService(final UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    public User findById(Long id) {
        return userRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("존재하지 않는 유저입니다. : " + id));
    }
}

```

dpudpu committed:
feat: CommentDto @JsonFormat(pattern = "yyyy-mm-dd HH:mm") 추가

woowahan-pjs reviewed (changes requested):
질문에 대한 답과 몇 가지의 추가 질문을 남겼습니다.
진행 과정에서 궁금한 점은 댓글이나 Slack을 통해 물어보세요.

dpudpu committed:
refactor: SpringExtension -> MockitoExtension 로 변경

dpudpu committed:
feat: 테스트용 샘플 데이터 추가

dpudpu committed:
refactor: CommentRestController Hateoas 일부 기능 적용, 리턴 방식 변경

dpudpu committed:
refactor: 패키지 구조 변경

dpudpu committed:
refactor: ArticleDto.Response 의존 변경 List<Comment> -> List<CommentDto.Response>

woowahan-pjs reviewed (approved):
블로그 미션을 수행하느라 대단히 고생 많으셨습니다. 💯 
좋은 개발자가 되기 위한 소중한 밑거름이 되었길 바랍니다.
이만 Merge 하겠습니다.
